modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlApiBaseTest.java

```diff
@@ -0,0 +1,860 @@
+        IgniteException e = assertThrows(IgniteException.class,
+                () -> checkDml(1, outerTx0, ses, "INSERT INTO TEST VALUES (?, ?)", ROW_COUNT, Integer.MAX_VALUE));
+        assertEquals(Transactions.TX_FAILED_READ_WRITE_OPERATION_ERR, e.code());
+
+        assertThrowsSqlException(
+                Sql.CONSTRAINT_VIOLATION_ERR,
+                "PK unique constraint is violated",
+                () -> checkDml(1, ses, "INSERT INTO TEST VALUES (?, ?)", ROW_COUNT, Integer.MAX_VALUE));
+
+        ResultSet<SqlRow> rs = executeForRead(ses, "SELECT VAL0 FROM TEST ORDER BY VAL0");
+
+        assertEquals(2 * ROW_COUNT, asStream(rs).count());
+
+        rs.close();
+
+        outerTx = igniteTx().begin();
+
+        rs = executeForRead(ses, outerTx, "SELECT VAL0 FROM TEST ORDER BY VAL0");
+
+        assertEquals(2 * ROW_COUNT, asStream(rs).count());
+
+        rs.close();
+
+        outerTx.commit();
+
+        checkDml(2 * ROW_COUNT, ses, "UPDATE TEST SET VAL0 = VAL0 + ?", 1);
+
+        checkDml(2 * ROW_COUNT, ses, "DELETE FROM TEST WHERE VAL0 >= 0");
+
+        assertEquals(ROW_COUNT + 1 + 1 + 1 + 1 + 1 + 1, txManagerInternal.finished() - txPrevCnt);
+
+        assertEquals(0, txManagerInternal.pending());
+    }
+
+    /** Check correctness of explicit transaction rollback. */
+    @Test
+    public void checkExplicitTxRollback() {
+        IgniteSql sql = igniteSql();
+
+        sql("CREATE TABLE TEST(ID INT PRIMARY KEY, VAL0 INT)");
+
+        Session ses = sql.createSession();
+
+        // Outer tx with further commit.
+        Transaction outerTx = igniteTx().begin();
+
+        for (int i = 0; i < ROW_COUNT; ++i) {
+            checkDml(1, outerTx, ses, "INSERT INTO TEST VALUES (?, ?)", i, i);
+        }
+
+        rollback(outerTx);
+
+        ResultSet<SqlRow> rs = executeForRead(ses, "SELECT VAL0 FROM TEST ORDER BY VAL0");
+
+        asStream(rs);
+        assertEquals(0, asStream(rs).count());
+
+        rs.close();
+    }
+
+    /** Check correctness of rw and ro transactions for table scan. */
+    @Test
+    public void checkMixedTransactionsForTable() {
+        sql("CREATE TABLE TEST(ID INT PRIMARY KEY, VAL0 INT)");
+
+        Matcher<String> planMatcher = containsTableScan("PUBLIC", "TEST");
+
+        checkMixedTransactions(planMatcher);
+    }
+
+
+    /** Check correctness of rw and ro transactions for index scan. */
+    @Test
+    public void checkMixedTransactionsForIndex() throws Exception {
+        sql("CREATE TABLE TEST(ID INT PRIMARY KEY, VAL0 INT)");
+        sql("CREATE INDEX TEST_IDX ON TEST(VAL0)");
+
+        Matcher<String> planMatcher = containsIndexScan("PUBLIC", "TEST", "TEST_IDX");
+
+        checkMixedTransactions(planMatcher);
+    }
+
+    private void checkMixedTransactions(Matcher<String> planMatcher) {
+        IgniteSql sql = igniteSql();
+
+        if (sql instanceof ClientSql) {
```

Comment:
I think it is better to move such guards into a test case body directly, so we know that the test is not expected to run for sql clients.